_posts/2019-06-06-NeuralODEs.markdown

```diff
@@ -0,0 +1,388 @@
+
+<br />
+
+# 4 Normalising Flows
+  **Motivation**: In this class we take a little detour through the topic of Normalising Flows. This is used for density estimation and generative modeling, and it is another model which can be seen a time-discretisation of its continuous-time counterpart. Understanding it is necessary to understand continuous normalizing flows, one of the applications of neural ODEs.
+
+  **Topics**:
+
+  1. Normalising Flows
+  2. End-to-end implementations with neural nets
+
+  **Required Reading**:
+
+  1. [Density Estimation by Dual Ascent of the Log-likelihood](https://math.nyu.edu/faculty/tabak/publications/CMSV8-1-10.pdf) (you can skip Section 3) (DE)
+  2. [A family of non-parametric density estimation algorithms](https://math.nyu.edu/faculty/tabak/publications/Tabak-Turner.pdf) 
+  3. [A post on Normalising flow](http://akosiorek.github.io/ml/2018/04/03/norm_flows.html)
+  
+  **Notes**: Here is a [link](/assets/nodes_notes/week4.pdf) to our notes for the lesson. 
+  
+  **Optional Reading**:
+
+  1. Variational Inference with Normalizing Flows](https://arxiv.org/pdf/1505.05770.pdf)
+  2. [High-Dimensional Probability Estimation with Deep Density Models](https://arxiv.org/pdf/1302.5125.pdf)
+  
+  **Questions**:
+
+  1. In (DE), what is the difference between \(rho_t\) and \(\tilde{\rho}_t\), i.e. what do they represent?
+     <details><summary>Solution</summary>
+     <p>
+     The function \(\tilde{\rho}_t\) is the density of the distribution of the random variable \(\phi_t^{-1}(y)\) where \(y\sim \mu\). The function \(\rho_t\) is the density of the distribution of the random variable \(\phi_t(x)\) where \(x\sim \rho\).
+     </p>
+     </details>
+
+  2. . What is the computational complexity of evaluating a determinant of a \(N\times N\) matrix, and why is that relevant in this context?
+     <details><summary>Solution</summary>
+     <p>
+     In general, the cost of computing the determinant of a \(N\times N\) matrix is \(O(N^3)\). To compute densities  transported by normalising flows, we need to compute the determinants of the Jacobians; therefore, an important feature of practical normalising flows, is that the Jacobian structure must allow an efficient computation of its determinant. See this week notes for more discussion on this.  
+     </p>
+     </details>
+
+<br />
+
+# 5 The adjoint method (and auto-diff)
+  **Motivation**: The adjoint method is a numerical method for efficiently computing the gradient of a function in numerical optimization problems. Understanding this method is essential to understand how to train ‘continuous depth’ nets. We also review the basics of Automatic Differentiation, which will help us understand the efficiency of the algorithm proposed in the NeuralODE paper. 
+
+  **Topics**:
+
+  1. Adjoint method
+  2. Auto-diff
+
+  **Required Reading**:
+
+  1. Section 8.7 from [Computational Science and Engineering](http://math.mit.edu/~gs/cse/) (CSE)
```

Comment:
Move (CSE) to the front like (DE) above.

Reply:
✅